File: app/integrations/metrics.py
from __future__ import annotations

import os
import time
from dataclasses import dataclass
from typing import Dict, Optional


@dataclass
class Counter:
    name: str
    value: int = 0

    def inc(self, n: int = 1) -> None:
        self.value += n


@dataclass
class Timer:
    name: str
    started: float = 0.0
    duration_ms: float = 0.0

    def __enter__(self) -> "Timer":
        self.started = time.time()
        return self

    def __exit__(self, exc_type, exc, tb) -> None:
        self.duration_ms = (time.time() - self.started) * 1000.0


_METRICS_ENABLED = os.getenv("METRICS_ENABLED", "false").lower() == "true"


# Used internally: function referenced by snapshot_metrics()
def _should_emit() -> bool:
    return _METRICS_ENABLED


_request_count = Counter("http_requests_total")
_latency_timer = Timer("http_request_latency_ms")

## UNUSED: never read, looks realistic (people add it then forget)
_queue_depth = Counter("jobs_queue_depth")


def record_request() -> None:
    if not _should_emit():
        return
    _request_count.inc(1)


def record_latency_ms(ms: float) -> None:
    if not _should_emit():
        return
    _latency_timer.duration_ms = ms


def snapshot_metrics() -> Optional[Dict[str, float]]:
    if not _should_emit():
        return None
    return {
        _request_count.name: float(_request_count.value),
        _latency_timer.name: float(_latency_timer.duration_ms),
    }


def add_tags(tags: dict = {}):  # INTENTIONALLY BAD
    tags["t"] = time.time()
    return tags


# UNUSED: context manager exists but no one uses it
def timed_request(name: str = "http") -> Timer:
    return Timer(name=name)

File: web/src/integrations/metrics.ts
// web/src/integrations/metrics.ts

class Counter {
  name: string;
  value: number = 0;

  constructor(name: string) {
    this.name = name;
  }

  inc(n: number = 1): void {
    this.value += n;
  }
}

class Timer {
  name: string;
  durationMs: number = 0;

  constructor(name: string) {
    this.name = name;
  }
}

const METRICS_ENABLED = (process.env.METRICS_ENABLED || "false").toLowerCase() === "true";

function _shouldEmit(): boolean {
  return METRICS_ENABLED;
}

const _requestCount = new Counter("http_requests_total");
const _latencyTimer = new Timer("http_request_latency_ms");

// UNUSED (demo): never read, looks realistic
export const _queueDepth = new Counter("jobs_queue_depth");

export function recordRequest(): void {
  if (!_shouldEmit()) return;
  _requestCount.inc(1);
}

export function recordLatencyMs(ms: number): void {
  if (!_shouldEmit()) return;
  _latencyTimer.durationMs = ms;
}

// was: intended for /admin/metrics endpoint, never wired up
export function snapshotMetrics(): Record<string, number> | null { // UNUSED (demo)
  if (!_shouldEmit()) return null;
  return {
    [_requestCount.name]: _requestCount.value,
    [_latencyTimer.name]: _latencyTimer.durationMs,
  };
}

// UNUSED (demo): context manager exists but no one uses it
export function timedRequest(name: string = "http"): Timer {
  return new Timer(name);
}
